### PR TITLE
Fix #150 - Fix colour editor for use in browsers that don't use -webkit- prefix

### LINKED ISSUE #150: Fix colour editor for use in browsers that don't use -webkit- prefix
The inline colour editor tool in Brackets isn't designed to work outside WebKit.  Here's what it looks like in the desktop Brackets app:

screen shot 2015-03-27 at 8 58 33 pm

Here's what it looks like in Bramble running in Firefox:

screen shot 2015-03-27 at 9 05 22 pm

Basically it just needs some tweaks to the LESS file used here https://github.com/adobe/brackets/blob/master/src/extensions/default/InlineColorEditor/css/main.less

I experimented a bit today and got to:

```diff
diff --git a/src/extensions/default/InlineColorEditor/css/main.less b/src/extensions/default/InlineColorEditor/css/main.less
index c9b46f3..a2d0d2a 100644
--- a/src/extensions/default/InlineColorEditor/css/main.less
+++ b/src/extensions/default/InlineColorEditor/css/main.less
@@ -55,7 +55,6 @@
     display: inline-block;
     position: relative;
     overflow: hidden;
-    -webkit-background-clip: padding-box;
     background-clip: padding-box;
     border-bottom: 1px solid rgba(255, 255, 255, 0.5);
 }
@@ -137,10 +136,27 @@
     float: left;
 }
 .color-editor section .color-selection-field .saturation-gradient {
-    background-image: -webkit-gradient(linear, 0% 50%, 100% 50%, from(#fff), to(rgba(255,255,255,0)));
+/*    background-image: -webkit-gradient(linear, 0% 50%, 100% 50%, from(#fff), to(rgba(255,255,255,0))); */
+
+
+background-image: -moz-linear-gradient(top,  rgba(255,255,255,1) 0%, rgba(255,255,255,0) 100%); /* FF3.6+ */
+background-image: -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(255,255,255,1)), color-stop(100%,rgba(255,255,255,0))); /* Chrome,Safari4+ */
+background-image: -webkit-linear-gradient(top,  rgba(255,255,255,1) 0%,rgba(255,255,255,0) 100%); /* Chrome10+,Safari5.1+ */
+background-image: -o-linear-gradient(top,  rgba(255,255,255,1) 0%,rgba(255,255,255,0) 100%); /* Opera 11.10+ */
+background-image: -ms-linear-gradient(top,  rgba(255,255,255,1) 0%,rgba(255,255,255,0) 100%); /* IE10+ */
+background-image: linear-gradient(to bottom,  rgba(255,255,255,1) 0%,rgba(255,255,255,0) 100%); /* W3C */
+
+
 }
 .color-editor section .color-selection-field .luminosity-gradient {
-    background-image: -webkit-gradient(linear, 0% 0%, 0% 100%, from(transparent), to(#000));
+/*    background-image: gradient(linear, 0% 0%, 0% 100%, from(transparent), to(#000)); */
+background-image: -moz-linear-gradient(top,  rgba(0,0,0,0) 0%, rgba(0,0,0,1) 100%); /* FF3.6+ */
+background-image: -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(0,0,0,0)), color-stop(100%,rgba(0,0,0,1))); /* Chrome,Safari4+ */
+background-image: -webkit-linear-gradient(top,  rgba(0,0,0,0) 0%,rgba(0,0,0,1) 100%); /* Chrome10+,Safari5.1+ */
+background-image: -o-linear-gradient(top,  rgba(0,0,0,0) 0%,rgba(0,0,0,1) 100%); /* Opera 11.10+ */
+background-image: -ms-linear-gradient(top,  rgba(0,0,0,0) 0%,rgba(0,0,0,1) 100%); /* IE10+ */
+background-image: linear-gradient(to bottom,  rgba(0,0,0,0) 0%,rgba(0,0,0,1) 100%); /* W3C */
+
 }
 .color-editor section .color-selection-field .selector-base {
     width: 12px;
@@ -202,7 +218,15 @@
     margin-right: 0px;
 }
 .color-editor section .hue-slider {
-    background-image: -webkit-linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00);
+/*    background-image: linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00); */
+background-image: rgb(255,0,0); /* Old browsers */
+background-image: -moz-linear-gradient(top,  rgba(255,0,0,1) 0%, rgba(255,0,255,1) 17%, rgba(0,0,255,1) 33%, rgba(0,255,255,1) 50%, rgba(0,255,0,1) 67%, rgba(255,255,0,1) 83%, rgba(255,0,0,1) 100%); /* FF3.6+ */
+background-image: -webkit-gradient(linear, left top, left bottom, color-stop(0%,rgba(255,0,0,1)), color-stop(17%,rgba(255,0,255,1)), color-stop(33%,rgba(0,0,255,1)), color-stop(50%,rgba(0,255,255,1)), color-stop(67%,rgba(0,255,0,1)), color-stop(83%,rgba(255,255,0,1)), color-stop(100%,rgba(255,0,0,1))); /* Chrome,Safari4+ */
+background-image: -webkit-linear-gradient(top,  rgba(255,0,0,1) 0%,rgba(255,0,255,1) 17%,rgba(0,0,255,1) 33%,rgba(0,255,255,1) 50%,rgba(0,255,0,1) 67%,rgba(255,255,0,1) 83%,rgba(255,0,0,1) 100%); /* Chrome10+,Safari5.1+ */
+background-image: -o-linear-gradient(top,  rgba(255,0,0,1) 0%,rgba(255,0,255,1) 17%,rgba(0,0,255,1) 33%,rgba(0,255,255,1) 50%,rgba(0,255,0,1) 67%,rgba(255,255,0,1) 83%,rgba(255,0,0,1) 100%); /* Opera 11.10+ */
+background-image: -ms-linear-gradient(top,  rgba(255,0,0,1) 0%,rgba(255,0,255,1) 17%,rgba(0,0,255,1) 33%,rgba(0,255,255,1) 50%,rgba(0,255,0,1) 67%,rgba(255,255,0,1) 83%,rgba(255,0,0,1) 100%); /* IE10+ */
+background-image: linear-gradient(to bottom,  rgba(255,0,0,1) 0%,rgba(255,0,255,1) 17%,rgba(0,0,255,1) 33%,rgba(0,255,255,1) 50%,rgba(0,255,0,1) 67%,rgba(255,255,0,1) 83%,rgba(255,0,0,1) 100%); /* W3C */
+
 }
 .color-editor section footer {
     font-size: 100%;
@@ -267,7 +291,7 @@
     font-size: 100%;
     vertical-align: middle;
     background: transparent;
-    -webkit-font-smoothing: antialiased;
+    font-smoothing: antialiased;
 }
 .color-editor ul.button-bar {
```

Which isn't bad, and looks like this:

screen shot 2015-03-27 at 9 02 36 pm

But it could be better.  And, since my goal is to get this landed upstream, I really want something that is exactly like the original--I just don't speak gradient very fluently.

Maybe I can convince @Pomax to take a crack at this?

### PR DESCRIPTION
**Pull request by AJDBenner**
Fix #150. This is what the fix looks like in Firefox.

screen shot 2015-04-01 at 1 43 14 pm

### CODE REVIEW

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 1]
```diff
@@ -199,10 +208,13 @@
 .color-editor section .slider.opacity-slider {
     background-image: url("../img/color_thumb_back.png");
     background-size: 8px;
-    margin-right: 0px;  
+    margin-right: 0px;
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: Revert this change, please.  Spurious whitespace-only change are to be avoided, since they make blame info harder to untangle.

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 2]
```diff
@@ -172,7 +172,10 @@ define(function (require, exports, module) {
         // Update gradients in color square & opacity slider
         this.$selectionBase.css("background-color", colorObject.toHexString());
         this.$opacityGradient.css("background-image", "-webkit-gradient(linear, 0% 0%, 0% 100%, from(" + hueColor + "), to(transparent))");
-        
+        this.$opacityGradient.css("background-image", "-moz-linear-gradient(top, " + hueColor +" 0%, rgba(0,0,0,0) 100%");
+        this.$opacityGradient.css("background-image", "-o-linear-gradient(top, " + hueColor +" 0%, rgba(0,0,0,0) 100%");
+        this.$opacityGradient.css("background-image", "-ms-linear-gradient(top, " + hueColor +" 0%, rgba(0,0,0,0) 100%");
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: We also need an unprefixed `linear-gradient` as the last option.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 3]
```diff
@@ -138,9 +141,15 @@
 }
 .color-editor section .color-selection-field .saturation-gradient {
     background-image: -webkit-gradient(linear, 0% 50%, 100% 50%, from(#fff), to(rgba(255,255,255,0)));
+    background-image: -moz-linear-gradient(left, #FFFFFF 0%, rgba(255,255,255,0) 100%);
+    background-image: -o-linear-gradient(left, #FFFFFF 0%, rgba(255,255,255,0) 100%);
+    background-image: -ms-linear-gradient(left, #FFFFFF 0%, rgba(255,255,255,0) 100%);
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: Same thing here, add a `linear-gradient` case too.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 4]
```diff
@@ -138,9 +141,15 @@
...
 }
 .color-editor section .color-selection-field .luminosity-gradient {
     background-image: -webkit-gradient(linear, 0% 0%, 0% 100%, from(transparent), to(#000));
+    background-image: -moz-linear-gradient(bottom, #000000 0%, rgba(255,255,255,0) 100%);
+    background-image: -o-linear-gradient(bottom, #000000 0%, rgba(255,255,255,0) 100%);
+    background-image: -ms-linear-gradient(bottom, #000000 0%, rgba(255,255,255,0) 100%);
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: And here.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 5]
```diff
@@ -199,10 +208,13 @@
...
     background-image: url("../img/color_thumb_back.png");
     background-size: 8px;
-    margin-right: 0px;  
+    margin-right: 0px;
 }
 .color-editor section .hue-slider {
     background-image: -webkit-linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00);
+    background-image: -moz-linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00);
+    background-image: -o-linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00);
+    background-image: -ms-linear-gradient(top, #f00, #f0f, #00f, #0ff, #0f0, #ff0, #f00);
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: And here.

**Comment by humphd** [comment]
Please fix issues I've pointed out above. How does this look in IE9/10?  

@le717, any thoughts?  This change seems like something we should just do upstream, to be honest.  I can't imagine there will be a lot of resistence, since it won't affect the `-webkit-*` case at all.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 6]
```diff
@@ -268,6 +280,10 @@
     vertical-align: middle;
     background: transparent;
     -webkit-font-smoothing: antialiased;
+    -moz-font-smoothing: antialiased;
+    -o-font-smoothing: antialiased;
+    -ms-font-smoothing: antialiased;
+    font-smoothing: antialiased;
```
Comment (le717) [inline src/extensions/default/InlineColorEditor/css/main.less]: Is `font-smoothing` even a valid (read: non-WebKit only) property, and do all these prefixes even exist? The closest I can find is a `font-smooth` property from a very old (2002) CSS3 draft, and it has long been dropped from the spec.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 7]
```diff
@@ -56,6 +56,9 @@
     position: relative;
     overflow: hidden;
     -webkit-background-clip: padding-box;
+    -moz-background-clip: padding-box;
+    -o-background-clip: padding-box;
+    -ms-background-clip: padding-box;
```
Comment (le717) [inline src/extensions/default/InlineColorEditor/css/main.less]: Also double-check `background-clip` has these prefixes.

**Comment by le717** [comment]
@humphd Aside from the few notes, I see no harm in landing this upstream, especially since it does not effect us at all (if anything, it may be good as it switches us to use the standard gradient syntax). I'm not sure if an issue won't be raised about the prefixes, though.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 2 above)
Comment (le717) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: CEF 2721 supports `-webkit-gradient`, `-webkit-linear-gradient` and `linear-gradient`. Which one, then, will be applied here?

File: src/extensions/default/InlineColorEditor/css/main.less (hunk 6 above)
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: Good question.  We have https://developer.mozilla.org/en-US/docs/Web/CSS/font-smooth in Firefox, but it's not really recommended.  Probably we should just revert this section of the patch, and leave it unchanged.

File: src/extensions/default/InlineColorEditor/css/main.less (hunk 7 above)
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: http://caniuse.com/#search=background-clip and https://developer.mozilla.org/en-US/docs/Web/CSS/background-clip and http://shouldiprefix.com/#background-image-options all seem hopeful about not needing prefixes on this.

File: src/extensions/default/InlineColorEditor/ColorEditor.js (hunk 2 above)
Comment (humphd) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: We'd need some guidance here from upstream, since ordering will matter here.

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 8]
```diff
@@ -88,7 +88,7 @@ define(function (require, exports, module) {
         this.$opacitySlider = this.$element.find(".opacity-slider");
         this.$opacitySelector = this.$element.find(".opacity-slider .selector-base");
         this.$swatches = this.$element.find(".swatches");
-        
+        console.log("LOOK HERE", this.$opacityGradient.selector);
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Remove this, please.

File: src/extensions/default/InlineColorEditor/css/main.less [hunk 9]
```diff
@@ -55,7 +55,6 @@
     display: inline-block;
     position: relative;
     overflow: hidden;
-    -webkit-background-clip: padding-box;
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/css/main.less]: You can probably leave this in.

File: src/extensions/default/InlineColorEditor/ColorEditor.js [hunk 10]
```diff
@@ -88,7 +88,6 @@ define(function (require, exports, module) {
         this.$opacitySlider = this.$element.find(".opacity-slider");
         this.$opacitySelector = this.$element.find(".opacity-slider .selector-base");
         this.$swatches = this.$element.find(".swatches");
-
```
Comment (humphd) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Revert this change

**Comment by humphd** [comment]
You should collect screenshots of this running in all the browsers, and also test it in Desktop Brackets.  Then we should rebase against Brackets' master and do a PR upstream.

**Comment by le717** [comment]
> You should collect screenshots of this running in all the browsers, and also test it in Desktop Brackets. Then we should rebase against Brackets' master and do a PR upstream. 

Just be sure to note the question about ordering in https://github.com/humphd/brackets/pull/158#discussion_r27601425.

**Comment by AJDBenner** [comment]
I've compiled screenshots for all the browsers.

Chrome
chrome

Firefox
firefox

Internet Explorer 10
ie10

Opera
opera

Safari
safari

Desktop Brackets
desktopbrackets

**Comment by AJDBenner** [comment]
> Just be sure to note the question about ordering in #158 (comment).

@le717 on desktop Brackets linear-gradient is used, no prefixes.

**Comment by AJDBenner** [comment]
I filed this upstream https://github.com/adobe/brackets/issues/10853
